benchmarks/models.py

```diff
@@ -0,0 +1,77 @@
+import math
+
+import torch
+import torch.nn as nn
+
+
+class EmbeddingLayer(nn.Embedding):
+    def __init__(self, ntoken, ninp, initrange):
+        super().__init__(ntoken, ninp)
+        self.ninp = ninp
+        self.weight.data.uniform_(-initrange, initrange)
+
+    def forward(self, src):
+        return super().forward(src) * math.sqrt(self.ninp)
+
+
+class PositionalEncodingLayer(nn.Module):
+    def __init__(self, d_model, dropout=0.1, max_len=5000):
+        super(PositionalEncodingLayer, self).__init__()
+        self.dropout = nn.Dropout(p=dropout)
+
+        pe = torch.zeros(max_len, d_model)
+        position = torch.arange(0, max_len, dtype=torch.float).unsqueeze(1)
+        div_term = torch.exp(torch.arange(0, d_model, 2).float() * (-math.log(10000.0) / d_model))
+        pe[:, 0::2] = torch.sin(position * div_term)
+        pe[:, 1::2] = torch.cos(position * div_term)
+        pe = pe.unsqueeze(0).transpose(0, 1)
+        self.register_buffer("pe", pe)
+
+    def forward(self, x):
+        x = x + self.pe[: x.size(0), :]
+        return self.dropout(x)
+
+
+class TransformerDecoderLayer(nn.TransformerEncoderLayer):
+    """Though this class inherits from torch.nn.TransformerEncoderLayer,
+    it functions as a decoder in this model"""
+
+    def __init__(self, ninp, nhead, nhid, droupout):
+        super().__init__(ninp, nhead, nhid, droupout)
+        self.src_mask = None
+
+    def _generate_square_subsequent_mask(self, sz):
+        mask = (torch.triu(torch.ones(sz, sz)) == 1).transpose(0, 1)
+        mask = mask.float().masked_fill(mask == 0, float("-inf")).masked_fill(mask == 1, float(0.0))
+        return mask
+
+    def forward(self, src):
+        if self.src_mask is None or self.src_mask.size(0) != len(src):
```

Comment:
we should document what's src.shape. len(src) is src.size(0), which is batch dimension, right? I thought the mask should be generated on the sequence dimension, not the batch dimension?

also, regenerating the mask whenever len(src) changes can be expensive if len(src) changes all the time. It seems better to generate it once for biggest possible value and reuse a smaller portion of the tensor depending on len(src)?

Reply:
Good point. I think this benchmark code follows this [tutoria](https://pytorch.org/tutorials/beginner/transformer_tutorial.html#load-and-batch-data) very closely where the data is supposed to be batched such that seq_len is the first dimension and batch_size is the second dimension. However I think there is an issue with the way we are batching now such that the first dimension is the batch size. Note, that this batching is correct if we use real data and not synthetic so there is some discrepancy between the two input pipelines.

I am going to refactor the input pipeline such that both the real and synthetic data pipelines use similar intermediate steps.